src/librustc_trans/lib.rs

```diff
@@ -149,13 +149,16 @@ impl !Send for LlvmTransCrate {} // Llvm is on a per-thread basis
 impl !Sync for LlvmTransCrate {}
 
 impl LlvmTransCrate {
-    pub fn new(sess: &Session) -> Box<TransCrate> {
-        llvm_util::init(sess); // Make sure llvm is inited
+    pub fn new() -> Box<TransCrate> {
         box LlvmTransCrate(())
     }
 }
 
 impl TransCrate for LlvmTransCrate {
+    fn init(&self, sess: &Session) {
+        llvm_util::init(sess); // Make sure llvm is inited
```

Comment:
I think this is fine for now but in the future we might want to have a `Session` created by the time we create this trait object (we'll need at least the target spec loaded).

Reply:
Maybe let `__rustc_codegen_backend` take a `Option<&Session>` as argument and pass None when we are only using it for printing version info and stuff like that.

Reply:
No, I think we can always create a `Session`, there shouldn't really be a significant cost to it since we're already doing the argument parsing and we have to load the target specs anyway.

Reply:
Ah yeah so the two locations this was necessary (not taking a `Session` on construction) was when we print the version and when we print the LLVM passes. There's also one where we need to construct a list of all error codes but that's commented out for now...

I it'd be possible to have a `Session` here, but it'd be a "dummy session" in some cases because I don't think we want to generate an error for failure to parse command line arguments before we print the version. Either way I'd be fine though!

Reply:
We don't fully parse arguments before processing the version? Is it a separate codepath?

Reply:
@eddyb yeah at least for now the codepaths are pretty separate, I think it's *while* we're parsing arguments we'd delegate out to this.

It does mean that various print modes won't work with custom backends immediately but that's probably ok for now. Eventually we may be able to just require a couple of symbols to be defined and continue to require a `Session` for this.

Or a different trait object is returned perhaps!

Reply:
I don't know, it seems fragile - are you saying `rustc --target=...emscripten -vV` and `rustc -vV --target=...emscripten` would behave differently? We could certainly *defer* CLI errors without preventing the creation of a `Session`, and not print them in certain situations.

Reply:
Oh no both of those wouldn't use `--target` at all, the `-vV` is detected early. It does mean though that both of those would not print the version of the emscripten backend (or print passes from it).

Do you think this should block this landing? I figured this was esoteric enough that there's no need to fix so soon.

Reply:
Well this PR doesn't use target specs yet, so it's not necessary *for now*. But I am curious, because I was hoping for a straight-forward implementation, is the "printing the version skips checking the rest of the CLI arguments" guaranteed/relied upon?

Reply:
I think it basically would just require more refactoring. We'd have to, for example, initialize the `Session` in stages, first with a bunch of known options, then do some processing, then aftewards actually validate everything. AFAIK it's not really set up to do that yet.

Reply:
Can't we validate everything early and buffer the errors or something?

Reply:
I guess? Again though I didn't get the impression that the argument parsing was at all ready for this sort of refactoring, in the sense that it seemed outside the scope of this PR

Reply:
Yeah, I don't want to block *this* PR, this is more for future reference.